### ISSUE TITLE
SC2129 wrongly flagging correct construct

### ISSUE THREAD

**Issue by IanBlakeley** [koalaman/shellcheck#1883]
- Rule Id == SC2129
- My shellcheck version = 0.4.6 also see in online version 

The construct I am using seems to be correct from the wiki but I see
`In get_iplayer-pvr.sh line 247:
{
^-- SC2129: Consider using { cmd1; cmd2; } >> file instead of individual redirects.
`
ian@bilbo2:~/Documents/Scripts$ bash --version
GNU bash, version 4.4.20(1)-release (x86_64-pc-linux-gnu)
Copyright © 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html

code snippet
```
#Calculate runtime 0
res2="$(date +%s)"
dt="$(echo "$res2 - $res1" | bc)"
dd="$(echo "$dt/86400" | bc)"
dt2="$(echo "$dt-86400*$dd" | bc)"
dh="$(echo "$dt2/3600" | bc)"
dt3="$(echo "$dt2-3600*$dh" | bc)"
dm="$(echo "$dt3/60" | bc)"
ds="$(echo "$dt3-60*$dm" | bc)"

{
echo "'date'"
echo
printf "Cache checked - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds"
} >> $playerlogfile
```

I think this should not trigger the warning or at least I am not sure why it does so. There are three similar warnings in the whole script but same printf to log is used elsewhere and does not trigger.

**Comment by Nightfirecat** [comment]
I don't see that error when copying your snippet into the online version.

**Comment by IanBlakeley** [comment]
Maybe my issue is coming from further up.

Here's the whole thing,


```
#!/bin/bash
PATH=/home/ian/bin:/usr/local/sbin:/usr/local/bin:/home/ian/.local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/ian/Documents/Scripts:/snap/bin:.:

#Uncomment the three below if debugging is needed
#set -x
#set -v
#trap "set +x; sleep .1; set -x" DEBUG



now=$(date +%H:%M)
runtime="07:15" # time we want our script to execute
runtime2="16:55" #second execution less 1 minute
holdtime="13:30" # do not execute twice if after this time
norepeat="900" # don't run again if run within last number of seconds
last_modified=$(stat -c %Y logs/get_iplayer-pvr.log)
current=$(date +%s)
lastrun=$((current - last_modified))
lastrundays=$((lastrun / 86400))
res1="$(date +%s)" # taking ages to run so adding to check the acutal runtimes
playerlogfile="/home/ian/Documents/Scripts/logs/get_iplayer-pvr.log"
errorfile="/home/ian/Documents/Scripts/logs/get_iplayer-pvr.error"
home="/home/ian/Documents/Scripts"
loopbreak=0

#Change to working directory
cd $home || exit

#VPS Server remove end of March 2020 no longer available
OVYOB="b8f246a5-ad82-4a04-a83c-6d09-42a6-97ca"

#UK VPN
VPN="b8f246a5-ad82-42a6-97ca-4602a9614b9b"

#Create the Locking dir if it doesn't exist
if [[ ! -d "lockdir/" ]]; then
    mkdir -p lockdir/
fi

#Check if there is currently a lock in place, if so then exit, if not then create a lock
if [ -f "lockdir/get_iplayer-pvr.lock" ]; then
	{
    echo "'date'"
    echo "get_iplayer-pvr is currently already running, lock file in place lockdir"
	echo
	} >> $errorfile
    exit
else
    touch lockdir/get_iplayer-pvr.lock
fi

#check if cache rebuild is ongoing if so wait until finished
for ((n=1;n<4;n++))
do 
if [ -f "lockdir/get_iplayer-cache.lock" ]; then
	{
    echo "'date'"
    echo "cache rebuild in progress, need to wait"
	echo
	} >> $errorfile
	sleep 600
	fi
done
if [ -f "lockdir/get_iplayer-cache.lock" ]; then 
exit	
else
    touch lockdir/get_iplayer-pvr.lock
fi

# reinstate lock 2 if pc started late in day, no need to run twice now can wait until 17:15
if [[ $now > $holdtime ]]
then
touch lockdir/get_iplayer-pvr2.lock
fi

file="lockdir/get_iplayer-pvr2.lock" 
		if [[ $now > $runtime2 ]] 
	then
		if [ -f $file ] ; then
		rm $file
		fi
	fi
	
file="lockdir/get_iplayer-pvr2.lock" 
		if [[ $now < $runtime ]] 
	then
		if [ -f $file ] ; then
		rm $file
		fi
	fi

# Check if there is a second lock in place, if so exit unless it is after 17:11 or before 08:15
if [ -f "lockdir/get_iplayer-pvr2.lock" ]; then
	{
   echo "'date'"
  echo "either get_iplayer-pvr has already run as lock2 file in place lockdir, or it is afternoon and not yet time to run again" 
	echo 
	} >> $errorfile
	#Also remove lock#1 or script will never run again
	rm lockdir/get_iplayer-pvr.lock
    exit
fi

#Check if we have run within the last 30 mins
if [ $lastrun -lt $norepeat ] ; then
  {
  echo "Script has just run, wait or  change value of norepeat"
  } >> $errorfile
  exit
fi

#remove oldest logs
file="logs/get_iplayer-pvr.bak2" 
if [ -f $file ] ; then
    rm $file
fi

#rename olderlogs
file="logs/get_iplayer-pvr.bak1"
if [ -f $file ] ; then
    mv $file "logs/get_iplayer-pvr.bak2"
fi
file="logs/get_iplayer-pvr.bak"
if [ -f $file ] ; then
    mv $file "logs/get_iplayer-pvr.bak1"
fi
file="logs/get_iplayer-pvr.log"
if [ -f $file ] ; then
    mv $file "logs/get_iplayer-pvr.bak"
fi

#list variables to log
  {
  echo "listing the current varibles" 
echo
echo "now = " "$now"
echo "runtime = " $runtime
echo "runtime2 = " $runtime2
echo "holdtime = " $holdtime
echo "norepeat = " $norepeat
echo "last_modified = " "$last_modified"
echo "current = " "$current"
echo "lastrun = " $lastrun
echo "lastrundays = " $lastrundays
echo "res1 = " "$res1"
echo "playerlogfile = " $playerlogfile
echo "errorfile = " $errorfile
echo "loopbreak = " $loopbreak
echo "home = " $home
echo "VPN - VPS = " $OVYOB
echo "UK VPN = " $VPN
echo
  }  >> $playerlogfile

lastrun=$((lastrun - norepeat))
echo "Last run value - repeat value = " $lastrun " If not +ve then run again" >> $playerlogfile

 #Calculate runtime # logs
res2="$(date +%s)"
dt="$(echo "$res2 - $res1" | bc)"
dd="$(echo "$dt/86400" | bc)"
dt2="$(echo "$dt-86400*$dd" | bc)"
dh="$(echo "$dt2/3600" | bc)"
dt3="$(echo "$dt2-3600*$dh" | bc)"
dm="$(echo "$dt3/60" | bc)"
ds="$(echo "$dt3-60*$dm" | bc)"
{
echo
printf "logs swapped - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds"
echo
} >> $playerlogfile

#check time & do not run if before 07:15, covers if started early as
#script is run from restart on cron

while [ $loopbreak -le 0 ] #while counter is 0 then loop
do
		now=$(date +%H:%M) #current time
		echo "Inside the loop, time check" "$now" >> $playerlogfile
			if [[ $now > $runtime ]] #check if current time is after wanted runtime
				then
					touch lockdir/get_iplayer-pvr2.lock
					{
					echo "'date'"  
					echo "Run time arrived, executing now"
					echo
					} >> logs/get_iplayer-pvr.log				
				else
			 	# convert times to seconds to calculate time to wait
				time2run=$(( $(date -d "$runtime" "+%s") - $(date -d "$now" "+%s") ))
				# wait 1 minute longer to avoid the situation where we keep going around the loop for up to 59 seconds
				sleep $((time2run + 60))
			fi
			#runtime has arrived so next incrment the counter so we will exit the loop
			((loopbreak++))
done

#Calculate runtime # time check
res2="$(date +%s)"
dt="$(echo "$res2 - $res1" | bc)"
dd="$(echo "$dt/86400" | bc)"
dt2="$(echo "$dt-86400*$dd" | bc)"
dh="$(echo "$dt2/3600" | bc)"
dt3="$(echo "$dt2-3600*$dh" | bc)"
dm="$(echo "$dt3/60" | bc)"
ds="$(echo "$dt3-60*$dm" | bc)"

{
echo "original value" "$now"
now=$(date +%H:%M) #current time
echo "updated value" "$now"
printf "Times checked going to run now (discrepancy noticed at this point) - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds" 
} >> $playerlogfile

#Check last run, if more than 30 days will need to rebuild cache, set it to
#  rebuild after 28 or more just in case
if [ $lastrun -gt 2419200 ]; then 
    {
        echo
        echo "target file is too old, need to rebuild cache"
        echo
   } >> logs/get_iplayer-pvr.log
        get_iplayer --cache-rebuild --type=tv,radio >> $playerlogfile   
		touch logs/cachelastrun
else 
    {
        echo
        echo "target file is new enough, $lastrundays days rebuild not required"
    } >> $playerlogfile
fi

#Calculate runtime 0
res2="$(date +%s)"
dt="$(echo "$res2 - $res1" | bc)"
dd="$(echo "$dt/86400" | bc)"
dt2="$(echo "$dt-86400*$dd" | bc)"
dh="$(echo "$dt2/3600" | bc)"
dt3="$(echo "$dt2-3600*$dh" | bc)"
dm="$(echo "$dt3/60" | bc)"
ds="$(echo "$dt3-60*$dm" | bc)"

{
echo "'date'"
echo
printf "Cache checked - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds"
} >> $playerlogfile

# Checking permissions
nmcli general permissions >> $playerlogfile
echo >> $playerlogfile
# Check active user
whoami >> $playerlogfile
echo >> $playerlogfile
id  >> $playerlogfile
echo >> $playerlogfile
#connect to VPN
echo "sending VPN up" >> $playerlogfile
nmcli con up uuid "$VPN" >> $playerlogfile
sleep 30 # pausing to give it a chance to connect
echo >> $playerlogfile

#Test VPN active, if not quit #1
if nmcli c show --active | grep -q "vpn"
    then
          echo "'date'" >> $playerlogfile
          echo "VPN is active" >> $playerlogfile
    else
          echo "'date'" >> $errorfile
          echo "'date'" >> $playerlogfile
          echo "VPN is not actived, quiting" >> $errorfile
          echo "VPN is not actived, quiting" >> $playerlogfile
          rm lockdir/get_iplayer-pvr*
          exit
fi

#Calculate runtime 1
res2="$(date +%s)"
dt="$(echo "$res2 - $res1" | bc)"
dd="$(echo "$dt/86400" | bc)"
dt2="$(echo "$dt-86400*$dd" | bc)"
dh="$(echo "$dt2/3600" | bc)"
dt3="$(echo "$dt2-3600*$dh" | bc)"
dm="$(echo "$dt3/60" | bc)"
ds="$(echo "$dt3-60*$dm" | bc)"

#added temp want to see what the actual values are
{
echo "'date'"
echo
echo "StartTime = " "$res1"
echo "EndTime = " "$res2"
echo "Elapsed Time dt = " "$dt"
echo "dd  Days = " "$dd"
echo "dt2 = " "$dt2"
echo "dh Hours = " "$dh"
echo "d3 = " "$dt3"
echo "dm Minutes = " "$dm"
echo "ds Seconds = " "$ds"
echo
} >> $playerlogfile

{
echo "'date'"
printf "Iplayer VPN active, starting Iplayer run - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds" 
echo
} >> $playerlogfile

#run the actual command, tv shows first
{
echo "'date'"
echo "Check for Doctor Who"
get_iplayer --pvr "Doctor_Who"
} >> $playerlogfile

#Test VPN active, if not quit #2
if nmcli c show --active | grep -q "vpn"
    then
          echo "'date'" >> $playerlogfile
          echo "VPN is still active" >> $playerlogfile
    else
          echo "'date'" >> $errorfile
          echo "'date'" >> $playerlogfile
          echo "VPN is not actived, quiting" >> $errorfile
          echo "VPN is not actived, quiting" >> $playerlogfile
          exit
fi

{
echo "'date'"
echo "Check for Other TV"
get_iplayer --pvr "TV"
} >> $playerlogfile

#Test VPN active, if not quit #3
if nmcli c show --active | grep -q "vpn"
    then
          echo "'date'" >> $playerlogfile
          echo "VPN is  still active" >> $playerlogfile
    else
          echo "'date'" >> $errorfile
          echo "'date'" >> $playerlogfile
          echo "VPN is not actived, quiting" >> $errorfile
          echo "VPN is not actived, quiting" >> $playerlogfile
           exit
fi

{
echo "'date'"
echo "Check for Radio"
get_iplayer --type=radio --mark-downloaded "The Archers"  #Mark the archers as downloaded so I don't end up with the omnibus edition
get_iplayer --pvr "Radio"
} >> $playerlogfile


#Test VPN active, if not quit #4
if nmcli c show --active | grep -q "vpn"
    then
          echo "'date'" >> $playerlogfile
          echo "VPN is still active" >> $playerlogfile
    else
          echo "'date'" >> $errorfile
          date >> $playerlogfile
          echo "VPN is not actived, quiting" >> $errorfile
          echo "VPN is not actived, quiting" >> $playerlogfile
          exit
fi

now=$(date +%H:%M)
file="lockdir/get_iplayer-pvr2.lock" 
	if [[ $now > $runtime2 ]] 
	then
		if [ -f $file ] ; then
		rm $file
		fi
	fi

#finished, tidy up
rm lockdir/get_iplayer-pvr.lock
touch logs/cachelastrun
{
date
echo "Downloading is completed, now process files."
echo
} >> $playerlogfile

#Calculate runtime 2
res2="$(date +%s)"
dt="$(echo "$res2 - $res1" | bc)"
dd="$(echo "$dt/86400" | bc)"
dt2="$(echo "$dt-86400*$dd" | bc)"
dh="$(echo "$dt2/3600" | bc)"
dt3="$(echo "$dt2-3600*$dh" | bc)"
dm="$(echo "$dt3/60" | bc)"
ds="$(echo "$dt3-60*$dm" | bc)"

#added temp want to see what the actual values are
{
echo "'date'"
echo
echo "StartTime = " "$res1"
echo "EndTime = " "$res2"
echo "Elapsed Time dt = " "$dt"
echo "dd  Days = " "$dd"
echo "dt2 = " "$dt2"
echo "dh Hours = " "$dh"
echo "d3 = " "$dt3"
echo "dm Minutes = " "$dm"
echo "ds Seconds = " "$ds"
echo
} >> $playerlogfile

{
echo "'date'"
printf "Iplayer download finished - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds"
echo
} >> $playerlogfile

#disconnect to VPN
nmcli con down uuid "$VPN" >> $playerlogfile
sleep 30 # pausing to give it a chance to disconnect
echo >> $playerlogfile

#Check VPN is disconnected
if nmcli c show --active | grep -q "vpn"
    then
    {
          echo "'date'"
          echo "VPN is still active"
          echo "'date'" 
          echo "'date'" >> $errorfile
          echo "VPN is still actived" >> $errorfile
     } >> $playerlogfile
    else
    {
          echo "'date'" >> $playerlogfile
          echo "VPN is successfully deactived"
     } >> $playerlogfile
fi

cd "/home/ian/Desktop/Iplayer Recordings" || exit
#process any video
./mp4-mkv.sh >> $playerlogfile  2>&1
{
echo "'date'"
echo "The pvr run has completed, I am exiting."
echo
} >> $playerlogfile  2>&1

#Calculate runtime 3
res2="$(date +%s)"
dt="$(echo "$res2 - $res1" | bc)"
dd="$(echo "$dt/86400" | bc)"
dt2="$(echo "$dt-86400*$dd" | bc)"
dh="$(echo "$dt2/3600" | bc)"
dt3="$(echo "$dt2-3600*$dh" | bc)"
dm="$(echo "$dt3/60" | bc)"
ds="$(echo "$dt3-60*$dm" | bc)"

#added temp want to see what the actual values are
{
echo "'date'"
echo
echo "StartTime = " "$res1"
echo "EndTime = " "$res2"
echo "Elapsed Time dt = " "$dt"
echo "dd  Days = " "$dd"
echo "dt2 = " "$dt2"
echo "dh Hours = " "$dh"
echo "d3 = " "$dt3"
echo "dm Minutes = " "$dm"
echo "ds Seconds = " "$ds"
echo
printf "Iplayer run finished - Total Runtime (download+process time) : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds" 
} >> $playerlogfile

exit
```

blessed if I can see what is triggering it. It is not a problem for me I thought I'd found a possible issue in the checker which I could help resolve

**Comment by Nightfirecat** [comment]
On L238-256, you have:

```
{
echo "'date'"
echo
printf "Cache checked - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds"
} >> $playerlogfile

# Checking permissions
nmcli general permissions >> $playerlogfile
echo >> $playerlogfile
# Check active user
whoami >> $playerlogfile
echo >> $playerlogfile
id  >> $playerlogfile
echo >> $playerlogfile
#connect to VPN
echo "sending VPN up" >> $playerlogfile
nmcli con up uuid "$VPN" >> $playerlogfile
sleep 30 # pausing to give it a chance to connect
echo >> $playerlogfile
```

It is recommending that you write something like the following instead:

```
{
echo "'date'"
echo
printf "Cache checked - Runtime so far is : %d Days, %02d Hours, %02d Minutes, %02.2f Seconds\n" "$dd" "$dh" "$dm" "$ds"

# Checking permissions
nmcli general permissions
echo
# Check active user
whoami
echo
id
echo
#connect to VPN
echo "sending VPN up"
nmcli con up uuid "$VPN"
} >> $playerlogfile

sleep 30 # pausing to give it a chance to connect
echo >> $playerlogfile
```

The same is true for the other SC2129 cases in this script.

**Comment by IanBlakeley** [comment]
Thanks. Sorry I did not see it coming from further up
